Fix for bugz #1389165 - extended route validation breaks included templates

ramr opened the pull request:
@knobunc  @liggitt   PTAL  Thx
addressed the concerns you had in bug 1389165 [link](https://bugzilla.redhat.com/show_bug.cgi?id=1389165) as well - there were already existing tests that covered some of those but added a few more.

Also fixes #11767

liggitt commented:
I'm not seeing where an expired or not-yet-valid cert is tolerated

liggitt commented on pkg/route/api/validation/validation_test.go:
```diff
@@ -987,6 +1011,166 @@ func TestExtendedValidateRoute(t *testing.T) {
+					Host: "www.in.the.future.test",
+					TLS: &api.TLSConfig{
+						Termination: api.TLSTerminationEdge,
+						Certificate: testValidInFutureCert,
```
it looks like this is only passing because you omit a CA, so no validation is done at all. need to have a test with CA with not-yet-valid and expired dates to make sure CA validation passes even when the cert timestamps aren't valid.

ramr replied:
ok - will add a few more tests with an CA/intermediate CA - the one above is with self-signed cert.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -125,18 +125,18 @@ func ExtendedValidateRoute(route *routeapi.Route) field.ErrorList {
+			DNSName: hostname,
+			Roots:   certPool,
+		}
 	}
```
when verifying, probably need to override the current time in VerifyOptions to be the NotAfter time on the cert

liggitt replied:
and I would add two tests with cert->intermediate->root chains:
* include the intermediate cert in the Certificate data
* include the intermediate cert in the CACertificate data

(examples of intermediate chains at https://github.com/openshift/origin/tree/master/pkg/auth/authenticator/request/x509request/testdata)

ramr replied:
intermediates would have to be part of the CA cert. will add the above tests and check. thx

liggitt commented:
I'm also unsure we should include the full value of the cert/key/ca field in validation errors... in practice, it makes display of those errors unusable

ramr commented:
@liggitt  can clean up the errors if we want - it follows the other errors which pass the field value to   `field.Invalid`.

liggitt commented:
yeah, in places where the data is confidential (like the key) or known to not be useful to display, we omit it and put a placeholder in like `"<certificate data>"`

ramr commented:
@liggitt  addressed your comments - added more tests for intermediate CAs + future valid/expired certs plus made changes to pass the intermediates options to the verifier and cleaned up the errors to not leak cert/key data.   PTAL Thx

One thing that was weird was for certs that expired a year or two back, setting the `CurrentTime` for the VerifyOptions (https://golang.org/pkg/crypto/x509/#VerifyOptions) works but if the certs expired in `2000` (one of the test data set has that), it still returns an expired cert error. So I did have to do an explicit check/guard for that and only return the error if: 
`if invalidErr, ok := err.(x509.CertificateInvalidError); !ok || invalidErr.Reason != x509.Expired { ...`.

ramr commented:
[test]

liggitt commented on pkg/route/api/validation/validation_test.go:
```diff
@@ -110,6 +110,30 @@ bGvtpjWA4r9WASIDPFsxk/cDEEEO6iPxgMOf5MdpQC2y2MU0rzF/Gg==
 
 	testDestinationCACertificate = testCACertificate
+
+	testValidInFutureCert = "-----BEGIN CERTIFICATE-----\nMIIDjzCCAnegAwIBAgIJAJcdKWMFNUEGMA0GCSqGSIb3DQEBCwUAMF0xCzAJBgNV\nBAYTAlVTMQswCQYDVQQIDAJDQTERMA8GA1UECgwIU2VjdXJpdHkxEzARBgNVBAsM\nClNlbGZTaWduZXIxGTAXBgNVBAMMEHNlbGYuc2lnbmVyLnRlc3QwIBcNMzgwMTAx\nMDgwMDAyWhgPMjA1ODAxMDEwODAwMDJaMF0xCzAJBgNVBAYTAlVTMQswCQYDVQQI\nDAJDQTERMA8GA1UECgwIU2VjdXJpdHkxEzARBgNVBAsMClNlbGZTaWduZXIxGTAX\nBgNVBAMMEHNlbG ...
```
in the future, for readibility, use raw strings and literal newlines:
```
testValidInFutureCert = `-----BEGIN CERTIFICATE-----
MIIDjzCCAnegAwIBAgIJAJcdKWMFNUEGMA0GCSqGSIb3DQEBCwUAMF0xCzAJBgNV
BAYTAlVTMQswCQYDVQQIDAJDQTERMA8GA1UECgwIU2VjdXJpdHkxEzARBgNVBAsM
ClNlbGZTaWduZXIxGTAXBgNVBAMMEHNlbGYuc2lnbmVyLnRlc3QwIBcNMzgwMTAx
...
```

knobunc commented:
@liggitt do you want that chance to be made now?  Or are we good to merge?

ramr commented:
Done - made the readability changes.
[test]

knobunc commented:
[merge]

liggitt commented:
question about date validation and needs squash

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -300,9 +301,14 @@ func validateCertificatePEM(certPEM string, options *x509.VerifyOptions) (*x509.
 	if options != nil {
+		// Ensure we don't report errors for expired certs or if
+		// the validity is in the future.
+		options.CurrentTime = cert.NotAfter.AddDate(0, 0, -1)
```
looks like the intermediate certs are failing... if we have a cert valid from A-B signed by another cert valid from C-D, and A-B/C-D are non-overlapping, that chain will not be valid at any single instant.

liggitt replied:
I'm wondering if we can just validate the well-formedness of the CA bundle and skip the validation.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -300,9 +301,14 @@ func validateCertificatePEM(certPEM string, options *x509.VerifyOptions) (*x509.
 	}
```
we probably want to make sure all the `CERTIFICATE` blocks in the PEM parse

liggitt replied:
not sure what to do if there are non-CERTIFICATE PEM blocks in here... will that choke the router?

ramr replied:
We are appending only the certs from the PEM to the cert pool. 
As re: router should be fine - only cert and key blocks would be used.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -125,23 +125,24 @@ func ExtendedValidateRoute(route *routeapi.Route) field.ErrorList {
 	if len(tlsConfig.CACertificate) > 0 {
-		certPool = x509.NewCertPool()
+		certPool := x509.NewCertPool()
 		if ok := certPool.AppendCertsFromPEM([]byte(tlsConfig.CACertificate)); !ok {
```
AppendCertsFromPEM returns ok even if the CA bundle contains invalid certs, which could break the router. `pkg/cmd/util/CertificatesFromPEM()` will return an error if there are malformed certs, which is probably what you want.

ramr commented:
@liggitt   made changes as per your comments. 
[test]

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -125,23 +125,28 @@ func ExtendedValidateRoute(route *routeapi.Route) field.ErrorList {
-			result = append(result, field.Invalid(tlsFieldPath.Child("caCertificate"), tlsConfig.CACertificate, "failed to parse CA certificate"))
+		certPool := x509.NewCertPool()
+		if certs, err := cmdutil.CertificatesFromPEM([]byte(tlsConfig.CACertificate)); err != nil {
+			result = append(result, field.Invalid(tlsFieldPath.Child("caCertificate"), "<ca certificate data>", "failed to parse CA certificate"))
```
include the error? `"failed to parse CA certificate: %v", err`

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -152,14 +157,13 @@ func ExtendedValidateRoute(route *routeapi.Route) field.ErrorList {
-		if ok := roots.AppendCertsFromPEM([]byte(tlsConfig.DestinationCACertificate)); !ok {
-			result = append(result, field.Invalid(tlsFieldPath.Child("destinationCACertificate"), tlsConfig.DestinationCACertificate, "failed to parse destination CA certificate"))
+		if _, err := cmdutil.CertificatesFromPEM([]byte(tlsConfig.DestinationCACertificate)); err != nil {
+			result = append(result, field.Invalid(tlsFieldPath.Child("destinationCACertificate"), "<destination ca certificate data>", "failed to parse destination CA certificate"))
```
include error in message

ramr commented:
@liggitt  done - [test]

ramr commented on pkg/route/api/validation/validation.go:
```diff
@@ -276,35 +282,27 @@ func validateInsecureEdgeTerminationPolicy(tls *routeapi.TLSConfig, fldPath *fie
+		_, err = certs[0].Verify(*options)
 		if err != nil {
-			return cert, fmt.Errorf("error verifying certificate: %s", err.Error())
+			if invalidErr, ok := err.(x509.CertificateInvalidError); !ok || invalidErr.Reason != x509.Expired {
```
The original issue was to ensure cert+key+ca matched/was valid - which is why the extended validation was added. So ... we would need this.

liggitt replied:
Are we sure the expiration error doesn't short circuit the signing validation error

ramr replied:
There's tests for invalid cacert/cert/key combinations with good certificates which test for that.

liggitt replied:
ok

liggitt replied:
add a comment about why we ignore expired certs (doesn't choke the router, shouldn't reject routes), and why we don't simulate the current time in verify options (can't pick a single point in time that validates if the cert was signed by an expired intermediate)

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -276,35 +282,27 @@ func validateInsecureEdgeTerminationPolicy(tls *routeapi.TLSConfig, fldPath *fie
-		_, err = cert.Verify(*options)
+		// Ensure we don't report errors for expired certs or if
+		// the validity is in the future.
+		options.CurrentTime = certs[0].NotAfter.AddDate(0, 0, -1)
```
if we're ignoring expired certs, there's not need to mess with CurrentTime at all

liggitt commented:
remove the currenttime hack, comment why we're ignoring expired certs, squash, then LGTM

ramr commented:
done - changed as per @liggitt  comments and squashed.
[test]

liggitt commented:
Thanks, LGTM

knobunc commented:
[merge]
